### DISCUSSION TITLE
ImageMagick center image group on canvas

### DISCUSSION THREAD

**Discussion by xtiaan3** [ImageMagick/ImageMagick#5603] [Help] [answered]
I am outputting a banner with an image to the left and some text to the right. This is working well, however the positioning is very manual. I'd like to be able to group the $robot_image and the $robot_name text into one image, and then center that image group on the background 1400x500 canvas. Any ideas how I'd achieve this? The attached image is what I'd like to achieve: The robot + text centered as a group on the canvas

```
$banner = exec("convert {$robot_image} -resize 500x500 -gravity west -geometry +100+100 -background '#{$hex}' -extent 1400x500 -font 'SignPainter.ttc' -pointsize 300 -fill 'white' -strokewidth 20 -fill white \
-stroke 'rgba(0,0,0,0.3)' -annotate +550+40 '{$robot_name}' \
-stroke none -annotate +550+40 '{$robot_name}' output.jpg");
```
![image](https://user-images.githubusercontent.com/16021630/192920661-746838ba-8697-45e3-9247-2fbf82195d1a.jpeg)

**Comment by fmw42** [comment]
Post your original images and your hex value. You can create a text image using "label:" from the text and append it to the right of your robot image. Then composite that over your background or extend the background color centered. -extext respect the -gravity setting.  See docs on -extent and label:

**Reply by xtiaan3** [reply]
Hello - TY! Attached is the original robot image. The hex value is dynamic, but in this case is #485d6e. I have tried what you suggested, still no luck. Could you show me a command line example?
![image](https://user-images.githubusercontent.com/16021630/193127497-4a2a7b0d-e70e-4e71-9f56-29d23ca2c737.png)

**Comment by fmw42** [comment]
Does this do what you want?

`convert robot.png -resize 500x500 -background "#31BEFF" -size 500x -font /Users/fred/library/fonts/SignPainter.ttf -gravity center label:"Blazer" -alpha off +append -gravity center -background "#31BEFF" -extent 1400x500 result.jpg`


![result](https://user-images.githubusercontent.com/13458218/193134055-6c3532f4-c19c-42f0-87ce-e5786465803e.jpg)

**Reply by xtiaan3** [reply]
Thats it! thank you! I have been at this a while. Just a quick question. Looks like "label" cuts off the top of this particular font. Any ideas how to add padding or space to the top to prevent this? Also, if I wanted to offset the text and shift it to the left a little - closer to the robot, how would I do that. Really appreciate your help 
![image](https://user-images.githubusercontent.com/16021630/193135728-df38e8ed-ab36-4470-ab45-af6dda0c29f7.jpeg)

**Comment by fmw42** [comment, marked as answer]
Replace +append with +smush X, where X is negative to make them overlap.  That moves them closer together.  As far as the cutoff, your font may not have good metrics. I do not see that with my copy.  So try downloading a different version of the font and see if that gets cut off.

You can specify the text with either -size or -pointsize.

`convert robot.png -resize 500x500 -background "#31BEFF" -size 600x -font /Users/fred/library/fonts/SignPainter.ttf -gravity center label:"Blazer123" -alpha off +smush -50 -gravity center -background "#31BEFF" -extent 1400x500 result.jpg`

![result](https://user-images.githubusercontent.com/13458218/193139826-adde04ca-d3a7-4987-af1b-7da96f4ef6d8.jpg)

**Reply by xtiaan3** [reply]
That is perfect - thanks very much for the help!

**Reply by xtiaan3** [reply]
@fmw42  - If I could just trouble you for one last thing then I can deploy this. I am trying to get a 20 point stroke on this text in rgba with a black opacity of 0.3, but have the foreground text remain 100% white. As you can see my the image, it fades the whole text image to 0.3 opacity. How would I get a 0.3 opacity stroke and keep the text at 100% white?

`$banner = exec("convert robot.png -resize 500x500 -background '#{$hex1}' -size 600x -font 'signpainter-2.ttf' -fill 'white' -gravity center -strokewidth 20 -fill 'white' -stroke 'rgba(0,0,0,0.3)' label:'Blazer12345' -alpha off +smush -50 -gravity center -background '#{$hex1}' -extent 1400x500 output.jpg`
![image](https://user-images.githubusercontent.com/16021630/193147124-05074bea-3310-42ae-8c76-645a40fe3dfe.jpeg)

**Comment by fmw42** [comment]
Works fine for me except I think you have your alpha and opacity swapped. In place of 0.3, use 0.7.  Also use a smaller strokewidth.

`convert robot.png -resize 500x500 -background "#31BEFF" -size 600x -font /Users/fred/library/fonts/SignPainter.ttf -gravity center -fill white -stroke "rgba(0,0,0,0.7)" -strokewidth 5 label:"Blazer123" -alpha off +smush -50 -gravity center -background "#31BEFF" -extent 1400x500 result.jpg`

![result](https://user-images.githubusercontent.com/13458218/193151373-08cb5cfd-d7a8-478e-a5e0-c3a49b62d64d.jpg)

**Reply by xtiaan3** [reply]
Im really not getting this 😬 Im trying to do 2 things now. I need that thick stroke, so Im trying to place a new label exactly over the first one only with no stroke, so the lower layer will have the thick stroke I need and the duplicate layer placed on top with no stroke should make it look correct - but its placing it to the right. Also, Im trying to mask out the original robots background color. This works fine if I don't include the label, but with the label code I get this black mask effect. Any ideas here. Again, really appreciate your help. Just trying to get this wrapped up.

`convert {$nft_art_path} -transparent '#{$hex1}' -resize 500x500 -background 'red' -size 700x -font \
    'signpainter-2.ttf' -fill 'white' -gravity center -strokewidth 20 \
    -stroke 'rgba(0,0,0,0.3)' label:'{$robot_name}' -alpha off +smush -10 \
    -stroke none label:'{$robot_name}' -alpha off +smush -10 \
    -gravity center -background 'red' -extent 1400x500 output.jpg`

![image](https://user-images.githubusercontent.com/16021630/193155061-d6398b80-63ea-470d-9baf-2ef4a41d01c9.jpeg)

**Reply by fmw42** [reply]
Please format your code with the <> button so it puts the line ending \ at the end of the lines.  It is to hard to read.  Your problem is that you should do the two strokes together and only one smush. But each label: will create its own image. So if you have two labels and an image, that will be three image that will append.  So you cannot do what you want that way.  You will have to create one image of your text using annotate, if you want to do multiple fonts on top of each other.  So create a background image of your color the size you want for the text or larger. Add your text with annotate, then trim to its bounding box. Then smush that with your image. Then extent to the final size you want.

**Reply by fmw42** [reply]
If you do not want the red zig-zag from the negative value of smush, then you need a different approach.  That would be:

Create a background image of your color. Then insert your image on the left. Then create your text image on transparent background or background the same color as the background.  Then use convert ... -composite to insert the text image where you desire near the image.  Then trim to the bounding box and extend to the final size you want.

**Reply by xtiaan3** [reply]
Thank you Fred. Im slowly learning the syntax. Would you be able to show me and example of the above command you mentioned. Im having trouble reproducing it. Once I get this im good to go

**Reply by fmw42** [reply]
What have you tried?

**Comment by fmw42** [comment]
I think this is what you want. I have to use annotate twice.  First draw your transparent color with stroke width, then draw your white over it.

Unix syntax

```
convert robot.png -resize 500x500 \
\( -size 700x700 xc:none \
-font /Users/fred/library/fonts/SignPainter.ttf -gravity center -pointsize 180 \
-fill "rgba(0,0,0,0.3)" -stroke "rgba(0,0,0,0.3)" -strokewidth 5 -annotate +0+0 "Blazer123" \
-fill "white" -annotate +0+0 "Blazer123" \
-trim +repage \) \
-background "#31BEFF" +smush -50 -extent 1400x500 result.jpg
```

![result](https://user-images.githubusercontent.com/13458218/193305810-3d8f6e1c-a7ed-40c0-8c68-ad8dd52997ce.jpg)

**Reply by xtiaan3** [reply]
Thank you Fred! I had to add a second -stroke none and that did the trick. Really appreciate all your help on this :) 

`convert {$nft_art_path} -transparent '#{$hex1}' -resize 500x500 \
\( -size 700x700 xc:none \
-font 'signpainter-2.ttf' -gravity center -pointsize 180 \
-fill 'rgba(0,0,0,0.3)' -stroke 'rgba(0,0,0,0.3)' -strokewidth 20 -annotate +0+0 'Christiaan' \
-fill 'white' -stroke none -annotate +0+0 'Christiaan' \
-trim +repage \) \
-background '#31BEFF' +smush 0 -extent 1400x500 output.jpg`

![image](https://user-images.githubusercontent.com/16021630/193319643-bf3e9498-4531-46b8-9549-7572b41d39de.jpeg)

**Reply by fmw42** [reply]
Good.  Looks nice.

**Reply by xtiaan3** [reply]
Fred if I could just trouble you for one last thing - I'd like to try and have an option to use an image as a background rather than a solid color. I replaced the -background color code with the image I'd like to use, but no luck. Im sure I have it out of order but can't seem too get it 

`convert {$nft_art_path} -transparent '#{$hex1}' -resize 500x500 \
    \( -size 700x700 xc:none \
    -font 'signpainter-2.ttf' -gravity center -pointsize {$font_size} \
    -fill 'rgba(0,0,0,0.3)' -stroke 'rgba(0,0,0,0.3)' -strokewidth {$stroke_size} -annotate +0+0 '{$robot_name}' \
    -fill 'white' -stroke none -annotate +0+0 '{$robot_name}' \
    -trim +repage \) \
    bgs/vintage_sun.jpg -resize 1400x500^ -gravity center -crop 1400x500 \
     -composite  output.jpg`

![vintage_sun](https://user-images.githubusercontent.com/16021630/193368184-2192e4dd-ce31-4533-bf02-467b54654a6b.jpg)

**Comment by fmw42** [comment]
```
convert robot.png -resize 500x500 \
\( -size 700x700 xc:none \
-font /Users/fred/library/fonts/SignPainter.ttf -gravity center -pointsize 180 \
-fill "rgba(0,0,0,0.3)" -stroke "rgba(0,0,0,0.3)" -strokewidth 5 -annotate +0+0 "Blazer123" \
-fill "white" -annotate +0+0 "Blazer123" \
-trim +repage \) \
-background none +smush -50 \
sunrise.jpg +swap -compose over -composite \
result.jpg
```

![result](https://user-images.githubusercontent.com/13458218/193368853-ec26ff31-8e6d-4aee-9c1e-bff1d343514f.jpg)

**Reply by xtiaan3** [reply]
worked perfect - Thank you again 🙏

![image](https://user-images.githubusercontent.com/16021630/193369396-fb646783-c450-4ea7-b6cb-48db649180e0.jpeg)
